midenc-compile/src/stages/cross_ctx/lower_imports.rs

```diff
@@ -0,0 +1,281 @@
+                import_func_sig,
+                analyses,
+                session,
+            )?;
+        }
+
+        let component_builder = component_builder.with_imports(lowered_imports);
+
+        let component = component_builder.build();
+
+        // dbg!(&component.imports());
+        // dbg!(&component.modules().keys());
+
+        Ok(LinkerInput::Hir(component.into()))
+    }
+}
+
+/// Generates the lowering function (cross-context Miden ABI -> Wasm CABI) for the given import function.
+fn generate_lowering_function(
+    component_builder: &mut ComponentBuilder<'_>,
+    high_func_ty: &FunctionType,
+    import_func_id: FunctionIdent,
+    import_func_sig: Signature,
+    span: SourceSpan,
+    diagnostics: &DiagnosticsHandler,
+) -> CompilerResult<(MidenAbiImport, FunctionIdent)> {
+    // get or create the module for the interface
+    let lowering_module_id =
+        Symbol::intern(format!("lower-imports-{}", import_func_id.module.as_str()));
+    // dbg!(&lowering_module_id);
+    let mut module_builder = component_builder.module(lowering_module_id);
+
+    let import_lowered_sig = flatten_function_type(high_func_ty, FlatteningDirection::Lower)
+        .map_err(|e| {
+            let message = format!(
+                "Miden CCABI import lowering generation. Signature for imported function {} \
+                 requires flattening. Error: {}",
+                import_func_id.function, e
+            );
+            diagnostics.diagnostic(Severity::Error).with_message(message).into_report()
+        })?;
+
+    if needs_transformation(&import_lowered_sig) {
+        let message = format!(
+            "Miden CCABI export lowering generation. Signature for exported function {} requires \
+             lowering. This is not supported yet.",
+            import_func_id
+        );
+        return Err(diagnostics.diagnostic(Severity::Error).with_message(message).into_report());
+    }
+    assert_core_wasm_signature_equivalence(&import_func_sig, &import_lowered_sig);
+
+    let mut builder = module_builder.function(import_func_id.function, import_func_sig)?;
+    let entry = builder.current_block();
+    let params = builder.block_params(entry).to_vec();
+
+    let dfg = builder.data_flow_graph_mut();
+    if dfg.get_import(&import_func_id).is_none() {
+        dfg.import_function(
+            import_func_id.module,
+            import_func_id.function,
+            import_lowered_sig.clone(),
+        )
+        .map_err(|_e| {
+            let message = format!(
+                "lowering function with name {} in module {} with signature \
+                 {import_lowered_sig:?} is already imported (function call) with a different \
+                 signature",
+                import_func_id.function, import_func_id.module
+            );
+            diagnostics.diagnostic(Severity::Error).with_message(message).into_report()
+        })?;
+    }
+    let call = builder.ins().call(import_func_id, &params, span);
```

Comment:
*Note:* This is where we will need to handle any of the `CanonLift` ABI details I mentioned in my other comment, both when calling and when handling the return values. I know you haven't implemented any of that yet, since we're only dealing with scalars at the moment, but I'd leave a comment/TODO to that effect here (and in the place I mentioned in my corresponding comment in the lifting code).

Reply:
Thanks! Done.